### ISSUE TITLE
GptOssAttention.init_weights leaves q/k/v/o projection biases uninitialized, causing NaN loss

### ISSUE THREAD

**Issue by janbernloehr** [NVIDIA-NeMo/Automodel#3920]
**Describe the bug**

`GptOssAttention.init_weights` initializes the weights of the `q_proj`, `k_proj`, `v_proj`, and `o_proj` linear layers, but it never initializes their biases.

All four projections are explicitly constructed with `bias=True`:

https://github.com/NVIDIA-NeMo/Automodel/blob/9a4632347323fbfee9ba1ae1a323669ef74fc5fb/nemo_automodel/components/models/gpt_oss/layers.py#L59-L70

However, `GptOssAttention.init_weights` only initializes `linear.weight`:

https://github.com/NVIDIA-NeMo/Automodel/blob/9a4632347323fbfee9ba1ae1a323669ef74fc5fb/nemo_automodel/components/models/gpt_oss/layers.py#L172-L187

```python
for linear in linear_list:
    nn.init.trunc_normal_(linear.weight, mean=0.0, std=init_std)
```

This is unsafe for the model's meta-device/materialization initialization path. Materializing parameters with `to_empty()` or `to_empty_parameters_only()` allocates storage without initializing its contents. Every parameter must therefore be written by the subsequent initializer.

The four attention projection biases are left with whatever happened to be present in the allocated storage. The contents may be zero by accident, but they may also be extremely large while still technically finite.

In a reproduction using the shipped GPT-OSS 20B benchmark configuration, one such omitted bias contained:

```text
model.layers.0.self_attn.q_proj.bias
numel=512
max_abs=1.70141e+38
non_finite=0
```

`1.70141e+38` is approximately `2^127`: a finite BF16 value, but far outside any plausible initialization range. Squaring values of this magnitude also explains why parameter-norm reporting can become `inf` even though an `isfinite()` check on the parameter itself succeeds.

The corrupted bias is added by `q_proj` before attention. In the observed failure:

- all inputs to `FlexAttention` were finite;
- the query magnitude reached approximately `2.29e+38`;
- `FlexAttention` was the first module to produce non-finite outputs;
- training reported `loss=nan` from iteration 0 onward.

This affects from-config/random initialization. A checkpoint load may subsequently overwrite the biases, but the shipped benchmark configuration constructs the model with `NeMoAutoModelForCausalLM.from_config`, so the incomplete initializer directly affects that workflow.

**Steps/Code to reproduce bug**

Reproduced on `main` at commit:

```text
9a4632347323fbfee9ba1ae1a323669ef74fc5fb
```

A deterministic unit reproduction can be added to `tests/unit_tests/models/gpt_oss/test_gptoss_layers.py`. The explicit NaN fill represents arbitrary storage returned by `to_empty()` and makes the otherwise allocator-dependent failure deterministic:

```python
def test_init_weights_initializes_projection_biases(gpt_config, backend_config):
    with torch.device("meta"):
        attention = GptOssAttention(gpt_config, backend_config)

    attention.to_empty(device="cpu")

    projections = [
        attention.q_proj,
        attention.k_proj,
        attention.v_proj,
        attention.o_proj,
    ]

    # Make the uninitialized-storage condition deterministic.
    with torch.no_grad():
        for projection in projections:
            assert projection.bias is not None
            projection.bias.fill_(float("nan"))

    attention.init_weights(torch.device("cpu"), init_std=0.02)

    for projection in projections:
        assert torch.isfinite(projection.bias).all()
        assert torch.count_nonzero(projection.bias).item() == 0
```

Run it with:

```bash
pytest -q tests/unit_tests/models/gpt_oss/test_gptoss_layers.py \
  -k initializes_projection_biases
```

On current `main`, the test fails because all four biases retain the injected NaNs. `init_weights()` rewrites the projection weights but never touches the biases.

The current GPT-OSS initializer test only verifies that `q_proj.weight` and `sinks` change. It does not inspect any projection bias:

https://github.com/NVIDIA-NeMo/Automodel/blob/9a4632347323fbfee9ba1ae1a323669ef74fc5fb/tests/unit_tests/models/gpt_oss/test_gptoss_layers.py#L194-L208

A production-level controlled comparison produced the following results. Both cases used the same GPT-OSS 20B configuration and environment; the only code delta was zero-initialization of the four projection biases.

| Observation | Current initializer | Biases zero-initialized |
| --- | --- | --- |
| Projection-bias audit | `layer.0.q_proj.bias` had `max_abs=1.70141e+38` | Every `q/k/v/o_proj.bias` had `max_abs=0` |
| Other model parameters | No other tensor exceeded `max_abs=6.46875` | No implausibly large tensors |
| First non-finite module | `FlexAttention`, despite finite inputs | None |
| Training result | All 60 emitted loss records across 30 iterations were `nan` | All 60 loss records were finite |
| Loss trajectory | `nan` from iteration 0 | `12.6999` → `12.2355` over 30 iterations |

Only `layer.0.q_proj.bias` happened to contain visibly bad data in that allocation. That does not narrow the source defect to one tensor: none of the four projection-bias families is written by `init_weights`, and uninitialized storage has arbitrary contents.

**Expected behavior**

After model materialization and `initialize_weights()`/`init_weights()`:

- every declared parameter should have a deterministic, valid value;
- the `q_proj`, `k_proj`, `v_proj`, and `o_proj` biases should be initialized;
- all parameters should be finite and within a plausible initialization range;
- the first forward and backward pass should remain finite;
- the shipped GPT-OSS 20B benchmark should not report NaN loss because of allocator contents.

The minimal fix is to zero-initialize each projection bias alongside its weight:

```python
for linear in linear_list:
    nn.init.trunc_normal_(linear.weight, mean=0.0, std=init_std)
    if linear.bias is not None:
        nn.init.zeros_(linear.bias)
```

Zero is consistent with the repository's handling of other model biases. For example, the shared MoE initializers explicitly zero gate and expert biases:

- https://github.com/NVIDIA-NeMo/Automodel/blob/9a4632347323fbfee9ba1ae1a323669ef74fc5fb/nemo_automodel/components/moe/layers.py#L928-L960
- https://github.com/NVIDIA-NeMo/Automodel/blob/9a4632347323fbfee9ba1ae1a323669ef74fc5fb/nemo_automodel/components/moe/experts.py#L1696-L1704

**Additional context**

### Source-level cause

The complete initialization chain does reach `GptOssAttention.init_weights`:

1. `GptOssForCausalLM.initialize_weights`
2. `GptOssModel.init_weights`
3. each transformer layer's `init_weights`
4. `self.self_attn.init_weights`

Relevant source:

- https://github.com/NVIDIA-NeMo/Automodel/blob/9a4632347323fbfee9ba1ae1a323669ef74fc5fb/nemo_automodel/components/models/gpt_oss/model.py#L86-L90
- https://github.com/NVIDIA-NeMo/Automodel/blob/9a4632347323fbfee9ba1ae1a323669ef74fc5fb/nemo_automodel/components/models/gpt_oss/model.py#L204-L218
- https://github.com/NVIDIA-NeMo/Automodel/blob/9a4632347323fbfee9ba1ae1a323669ef74fc5fb/nemo_automodel/components/models/gpt_oss/model.py#L354-L371

The problem is therefore not that the attention initializer is skipped. It is called, but it incompletely initializes the parameters owned by the attention module.

### Why this is independent of the loss implementation

The invalid value exists before the first training forward pass and before the loss function is evaluated. Changing between ordinary, chunked, or fused cross entropy cannot repair an invalid attention parameter.

PR #3909 correctly fixes the separate logits-memory issue by using `FusedLinearCrossEntropy`:

https://github.com/NVIDIA-NeMo/Automodel/pull/3909

That change does not touch `nemo_automodel/components/models/gpt_oss/layers.py`. It removes the earlier OOM and makes the initialization-induced NaN the next observable failure in the 20B benchmark.

The earlier issue already noted non-finite initialization and iteration-0 NaN as a possibly separate problem:

https://github.com/NVIDIA-NeMo/Automodel/issues/3907

### Suggested regression coverage

In addition to the deterministic per-module test above, it would be useful to add a GPT-OSS equivalent of the existing DeepSeek V4.1 meta-materialization test:

https://github.com/NVIDIA-NeMo/Automodel/blob/9a4632347323fbfee9ba1ae1a323669ef74fc5fb/tests/unit_tests/models/deepseek_v41/test_attention.py#L320-L330

The GPT-OSS regression should:

1. construct the attention layer or a small model on the `meta` device;
2. materialize it with the same empty-parameter path used by training;
3. invoke the model initializer;
4. verify every parameter is finite;
5. verify every projection bias is zero;
6. run a small forward and first backward pass;
7. verify outputs, loss, input gradients, and parameter gradients remain finite.

The deterministic sentinel/NaN test is still valuable because newly allocated memory can sometimes happen to contain zeros. A test that relies only on allocator contents may pass nondeterministically while the initializer remains incomplete.

### Scope

The missing initialization is shared by the linear projections used with both supported attention backends because it occurs before the Flex/TE attention implementation is called. The full runtime failure described above has been confirmed with the FlexAttention path; this report does not claim that the same visible failure has been reproduced with the TE attention path.

This also does not claim a last-known-good revision. The defect may have existed since the GPT-OSS implementation was introduced.
